src/adapter/src/coord/ddl.rs

```diff
@@ -85,6 +85,65 @@ impl Coordinator {
         self.catalog_transact_with(conn_id, ops, |_| Ok(())).await
     }
 
+    /// Executes a Catalog transaction with handling if the provided `Session` is in a SQL
+    /// transaction that is executing DDL.
+    pub(crate) async fn catalog_transact_with_ddl_transaction(
+        &mut self,
+        session: &mut Session,
+        ops: Vec<catalog::Op>,
+    ) -> Result<(), AdapterError> {
+        let conn_id = session.conn_id().clone();
+        let Some(Transaction {
+            ops:
+                TransactionOps::DDL {
+                    ops: txn_ops,
+                    revision: txn_revision,
+                    state: _,
+                },
+            ..
+        }) = session.transaction().inner()
+        else {
+            return self.catalog_transact(Some(session), ops).await;
+        };
+
+        // Make sure our Catalog hasn't changed since openning the transaction.
+        if self.catalog().transient_revision() != *txn_revision {
+            return Err(AdapterError::DDLTransactionRace);
+        }
+
+        // Combine the existing ops with the new ops so we can replay them.
+        let mut all_ops = Vec::with_capacity(ops.len() + txn_ops.len());
+        all_ops.extend(txn_ops.iter().cloned());
+        all_ops.extend(ops.clone());
+
+        // Run our Catalog transaction, but abort before committing.
+        let result = self
+            .catalog_transact_with(Some(&conn_id), all_ops.clone(), |state| {
+                // Return an error so we don't commit the Catalog Transaction, but include our
+                // updated state.
+                Err(AdapterError::DDLTransactionDryRun {
+                    new_ops: ops,
+                    new_state: state.catalog.clone(),
+                })
+            })
+            .await;
+
+        match result {
+            // We purposefully fail with this error to prevent committing the transaction.
+            Err(AdapterError::DDLTransactionDryRun { new_ops, new_state }) => {
+                // Adds these ops to our transaction, bailing if the Catalog has changed since we
+                // ran the transaction.
+                session.transaction_mut().add_ops(TransactionOps::DDL {
+                    ops: new_ops,
```

Comment:
```suggestion
                    ops: all_ops,
```

`new_state` is all of the ops applied, while `new_ops` is just the new ops. Should this instead be `all_ops` so that a `TransactionOps::DDL` is always consistent with itself (its state is always all of its ops)?

Reply:
Good thought, the consistency is key, updated!